internal/*: drop merging authorized_keys.d into authorized_keys

rfairley opened the pull request:
Will rebase this PR once #749 is merged.

So far I have verified that the blackbox test works, and that the old blackbox test fails with the build flag
`GLDFLAGS+="-X github.com/coreos/ignition/internal/distro.useAuthorizedKeysFile=true "` (in `./build_blackbox_tests`).

Still to do:
- [x] Verify in a built OS image that this works (with/without `useAuthorizedKeysFile=true`)
- [ ] Add documentation mentioned in #716 
- [x] Thinking of adding a blackbox test with the `distro.useAuthorizedKeysFile=true` flag that checks that `authorized_keys` gets updated

Fixes: #716

rfairley commented:
> Thinking of adding a blackbox test with the distro.useAuthorizedKeysFile=true flag that checks that authorized_keys gets updated

Hmm - finding it a bit tricky to add a test for this as writing to `authorized_keys` requires the linker flag setting which would need to be done in `./build_blackbox_tests`. I wonder about setting `distro.useAuthorizedKeysFile` to `"true"` at runtime before executing a blackbox test, but `distro.useAuthorizedKeysFile` needs to be exported for that.

I think since Ignition v3 configs support `authorized_keys.d` over `authorized_keys`,  it should be sufficient having the blackbox test here which has been adjusted to check for the fragment in `authorized_keys.d`.

rfairley commented:
Ready for a review

bgilbert commented on doc/configuration-v3_0-experimental.md:
```diff
@@ -104,7 +104,7 @@ The Ignition configuration is a JSON document conforming to the following specif
     * **name** (string): the username for the account.
     * **_passwordHash_** (string): the encrypted password for the account.
-    * **_sshAuthorizedKeys_** (list of strings): a list of SSH keys to be added to the user's authorized_keys.
+    * **_sshAuthorizedKeys_** (list of strings): a list of SSH keys to be added as an SSH key fragment at `authorized_keys.d/ignition` in the user's .ssh directory. To properly read the keys added by Ignition and allow the user to log in, the distribution should support reading key fragments from `authorized_keys.d`. For distributions that do not support `authorized_keys.d`, the keyfile written b ...
```
"`.ssh/authorized_keys.d/ignition` in the user's home directory" would be clearer.

I don't think distro integration should be mentioned in the spec.  It should go in the docs requested in #649.

rfairley replied:
Thanks - makes sense! Will have a think on the docs there.

bgilbert commented:
For the blackbox tests we could expose the option via an environment variable, the same as for `systemConfigDir`.

rfairley commented:
Made some edits:
- Dropped the code in `authorized_keys_d` and brought `as_user` to the directory level above (I think the helper functions in `as_user` will still be needed, for writing the keyfile with the right uid/gid).
- Made it so that if `.ssh/authorized_keys` is written, `.ssh/authorized_keys.d/ignition` is not. Would this behavior be preferred for distros that don't support `authorized_keys.d`? Thinking if merging doesn't need to be performed, then we can avoid having `authorized_keys.d` appear altogether.
- Made `distro.useAuthorizedKeysFile` configurable with an environment variable. I did wonder about having this option be a path instead, i.e. `distro.authorizedKeysFile`. I can't think of any use case where any path other than `authorized_key` or `authorized_keys.d/ignition` are needed, but having it a path would simplify the logic in `internal/exec/util/passwd.go` (could just pass the given path into `writeAuthKeysFile` rather than have a conditional).

To do:
- Add a blackbox test that uses `IGNITION_USE_AUTHORIZED_KEYS_FILE=true`. Will look into this tomorrow.

rfairley commented:
⬆ Pushed changes to add another blackbox test for `.ssh/authorized_keys` when `IGNITION_USE_AUTHORIZED_KEYS_FILE=true`. This adds `IgnitionEnv` as a member to `types.Test`. Also tidies up `passwd.go` a bit since removing `authorized_keys_d`.

Diff since the push yesterday: https://github.com/coreos/ignition/compare/6a6b24d5fffa0a957672c641a2361e730da11109..46f711b002e2ee7d988ccab6bac9ef1572477bef.

arithx commented:
ok to test

rfairley commented:
Lifted WIP - ready for another look.

For the distro-specific documentation part, should we handle that later in #649? (I left a comment there - so we won't forget). Or, I could get a doc started here in this PR, which #649 will add to/revise.

bgilbert commented:
I'm okay handling the additional docs in #649.

bgilbert commented on internal/distro/distro.go:
```diff
@@ -57,6 +57,10 @@ var (
+	// useAuthorizedKeysFile indicates whether to write SSH keys specified
+	// in the Ignition config to ".ssh/authorized_keys" ("true") or
+	// ".ssh/authorized_keys.d/ignition" ("false").
+	useAuthorizedKeysFile = "false"
```
nit: maybe `writeAuthorizedKeysFragment` (default true) would be clearer?

bgilbert commented on internal/distro/distro.go:
```diff
@@ -85,6 +89,9 @@ func XfsMkfsCmd() string   { return xfsMkfsCmd }
 func SelinuxRelabel() bool  { return bakedStringToBool(selinuxRelabel) }
 func BlackboxTesting() bool { return bakedStringToBool(blackboxTesting) }
+func UseAuthorizedKeysFile() bool {
+	return bakedStringToBool(fromEnv("USE_AUTHORIZED_KEYS_FILE", useAuthorizedKeysFile))
```
nit: Please put the body and declaration on the same line.

bgilbert commented on tests/positive/passwd/users.go:
```diff
@@ -140,21 +141,96 @@ ENCRYPT_METHOD SHA512
+				User:      1000,
+				Group:     1000,
+			},
+			Contents: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDBRZPFJNOvQRfokigTtl0IBi71LHZrFOk4EJ3Zowtk/bX5uIVai0Cd4+hqlocYL10idgtFBH28skeKfsmHwgS9XwOvP+g+kqAl7yCz8JEzIUzl1fxNZDToi0jA3B5MwXkpt+IWfnabwi2cRZhlzrz9rO+eExu5s3NfaRmmmCYrjCJIRPKSCrW8U0n9fVSbX4PDdMXVmH7r+t8MtR8523vCbakFR/Y0YIqkPVdfuUXHh9rDCdH4B7mt7nYX2LWQXGUvmI13mgQoy04ifkaR3ImuOMp3Y1J1gm6clO74IMCq/sK9+XJhbxMPPHUoUJ2EwbaG7Dbh3iqz47e9oVki4gIH step ...
```
If the Ignition binary is built with a non-default `useAuthorizedKeysFile` setting, won't this fail?  We should probably pass the env var in this test too.

rfairley replied:
:+1: - thanks for catching. Fixed this.

ajeddeloh commented on tests/positive/passwd/users.go:
```diff
@@ -21,12 +21,15 @@ import (
 	in := types.GetBaseDisk()
 	out := types.GetBaseDisk()
+	var env []string
+	env = append(env, "IGNITION_WRITE_AUTHORIZED_KEYS_FRAGMENT=true")
```
Super duper minor nit:
```
env := []string{"IGNITION...."}
```

rfairley replied:
Looks a lot cleaner, thanks!

ajeddeloh commented on tests/positive/passwd/users.go:
```diff
@@ -140,21 +143,96 @@ ENCRYPT_METHOD SHA512
+	in := types.GetBaseDisk()
+	out := types.GetBaseDisk()
+	var env []string
+	env = append(env, "IGNITION_WRITE_AUTHORIZED_KEYS_FRAGMENT=false")
```
same nit as above

rfairley commented:
Updated! I split the added test into its own commit - which also updates the existing `AddPasswdUsers` test to set the environment variable in the same commit.

rfairley commented:
Hmm - failure just seems to be from the gofmt. Thinking it wants to indent the function due to the line length. Any way to disable/configure gofmt to not consider that a failure? Having the one-line function body is neater. https://travis-ci.org/coreos/ignition/jobs/502719646#L603

ajeddeloh commented:
I'd say go with gofmt in this case, it is quite a long line otherwise.

rfairley commented:
Fixed! (had committed but forgot to squash it into 6b434dd just then). Should be good now.

bgilbert commented:
@rfairley Good to merge?

rfairley commented:
@bgilbert Good to merge!
